src/test/regress/expected/multi_view.out

```diff
@@ -251,6 +251,75 @@ SELECT * FROM mode_counts WHERE l_shipmode = 'AIR' ORDER BY 2 DESC, 1 LIMIT 10;
 (1 row)
 
 DROP MATERIALIZED VIEW mode_counts;
+-- modify statements on/with views
+-- create two tables and a view
+CREATE TABLE large (id int, tenant_id int);
+CREATE TABLE small (id int, tenant_id int);
+SELECT create_distributed_table('large','tenant_id');
+ create_distributed_table 
+--------------------------
+ 
+(1 row)
+
+SELECT create_distributed_table('small','tenant_id');
+ create_distributed_table 
+--------------------------
+ 
+(1 row)
+
+CREATE VIEW small_view AS SELECT * from small where id < 100;
+\copy small FROM STDIN DELIMITER ','
+\copy large FROM STDIN DELIMITER ','
+-- running modify statements "on" views still is not supported, hence below two statements will fail
+UPDATE small_view SET id = 1;
+ERROR:  table name "small" specified more than once
```

Comment:
This error is not expected. We get the error from the from the workers because we send the following query:
```
 DELETE FROM public.small_102330 small USING public.small_102330 small WHERE (small.id OPERATOR(pg_catalog.<) 100)

```

I think the issue is that we're now calling `ExtractRangeTableRelationWalker `, which only pulls the relations. But, in  `ModifyQuerySupported` , we check for `if (rangeTableEntry->relkind == RELKIND_VIEW)`, which is not in the list.

We should fix this

Reply:
Note that similarly we're losing all the coverage for the `else` part of `if (rangeTableEntry->rtekind == RTE_RELATION)` in [here](https://github.com/citusdata/citus/blob/870b743b938121b52e0967e14bf41347f1926d54/src/backend/distributed/planner/multi_router_planner.c#L691-L751)

Reply:
Would be good to improve this error message. We may be able to detect this case in originalQuery.

Reply:
I attempted to fix the bad error message issue in e844b596e45fed88fcef37fd64540adddaadbcb3

Reply:
Update: I guess I found a better way to solve it, I will try on it soon

Reply:
I tried to check if `resultRelation`th entry  of `rtable`  field of  `originalQuery/ ModifyQuerySupported` is a view or not. However, I guess it does not correspond to the query before rewrite process of postgress. This is because, `ModifyQuerySupported` is called via `distributed_planner` hook,  which is run at function call at `exec_simple_query:1143`. However, postgres's query rewrite is run at `exec_simple_query:1140`. After query rewrite, postgres appends the relation that the view refers to into the `rtable` and sets `resultRelation` to the index of that referred relation.

It would be better if we had the original query (the query itself before the rewrite) hence we can deduce if the modify statement is run on a view or not.

Reply:
For `update small_view set id=1` statement:

Before postgres rewrite, we have:
```
   {QUERY 
   :commandType 2 
   :querySource 0 
   :canSetTag true 
   :utilityStmt <> 
   :resultRelation 1 
   :hasAggs false 
   :hasWindowFuncs false 
   :hasTargetSRFs false 
   :hasSubLinks false 
   :hasDistinctOn false 
   :hasRecursive false 
   :hasModifyingCTE false 
   :hasForUpdate false 
   :hasRowSecurity false 
   :cteList <> 
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname small_view 
         :colnames ("id" "tenant_id")
         }
      :rtekind 0 
      :relid 16774 
      :relkind v 
      :rellockmode 3 
      :tablesample <> 
      :lateral false 
      :inh true 
      :inFromCl false 
      :requiredPerms 4 
      :checkAsUser 0 
      :selectedCols (b)
      :insertedCols (b)
      :updatedCols (b 8)
      :extraUpdatedCols (b)
      :securityQuals <>
      }
   )
   :jointree 
      {FROMEXPR 
      :fromlist (
         {RANGETBLREF 
         :rtindex 1
         }
      )
      :quals <>
      }
   :targetList (
      {TARGETENTRY 
      :expr 
         {CONST 
         :consttype 23 
         :consttypmod -1 
         :constcollid 0 
         :constlen 4 
         :constbyval true 
         :constisnull false 
         :location 25 
         :constvalue 4 [ 1 0 0 0 0 0 0 0 ]
         }
      :resno 1 
      :resname id 
      :ressortgroupref 0 
      :resorigtbl 0 
      :resorigcol 0 
      :resjunk false
      }
   )
   :override 0 
   :onConflict <> 
   :returningList <> 
   :groupClause <> 
   :groupingSets <> 
   :havingQual <> 
   :windowClause <> 
   :distinctClause <> 
   :sortClause <> 
   :limitOffset <> 
   :limitCount <> 
   :rowMarks <> 
   :setOperations <> 
   :constraintDeps <> 
   :withCheckOptions <> 
   :stmt_location 0 
   :stmt_len 26
   }
```

.After postgres query rewrite:

`originalQuery/ModifyQuerySupported` is 

```
   {QUERY 
   :commandType 2 
   :querySource 0 
   :canSetTag true 
   :utilityStmt <> 
   :resultRelation 2 
   :hasAggs false 
   :hasWindowFuncs false 
   :hasTargetSRFs false 
   :hasSubLinks false 
   :hasDistinctOn false 
   :hasRecursive false 
   :hasModifyingCTE false 
   :hasForUpdate false 
   :hasRowSecurity false 
   :cteList <> 
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname small_view 
         :colnames ("id" "tenant_id")
         }
      :rtekind 0 
      :relid 16774 
      :relkind v 
      :rellockmode 3 
      :tablesample <> 
      :lateral false 
      :inh true 
      :inFromCl false 
      :requiredPerms 4 
      :checkAsUser 0 
      :selectedCols (b)
      :insertedCols (b)
      :updatedCols (b 8)
      :extraUpdatedCols (b)
      :securityQuals <>
      }
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname small 
         :colnames ("id" "tenant_id")
         }
      :rtekind 0 
      :relid 16769 
      :relkind r 
      :rellockmode 3 
      :tablesample <> 
      :lateral false 
      :inh true 
      :inFromCl true 
      :requiredPerms 4 
      :checkAsUser 10 
      :selectedCols (b 8 9)
      :insertedCols (b)
      :updatedCols (b 8)
      :extraUpdatedCols (b)
      :securityQuals <>
      }
   )
   :jointree 
      {FROMEXPR 
      :fromlist (
         {RANGETBLREF 
         :rtindex 2
         }
      )
      :quals <>
      }
   :targetList (
      {TARGETENTRY 
      :expr 
         {CONST 
         :consttype 23 
         :consttypmod -1 
         :constcollid 0 
         :constlen 4 
         :constbyval true 
         :constisnull false 
         :location 25 
         :constvalue 4 [ 1 0 0 0 0 0 0 0 ]
         }
      :resno 1 
      :resname id 
      :ressortgroupref 0 
      :resorigtbl 0 
      :resorigcol 0 
      :resjunk false
      }
   )
   :override 0 
   :onConflict <> 
   :returningList <> 
   :groupClause <> 
   :groupingSets <> 
   :havingQual <> 
   :windowClause <> 
   :distinctClause <> 
   :sortClause <> 
   :limitOffset <> 
   :limitCount <> 
   :rowMarks <> 
   :setOperations <> 
   :constraintDeps <> 
   :withCheckOptions <> 
   :stmt_location 0 
   :stmt_len 26
   }
``` 
and `queryTree/ModifyQuerySupported` is 

```
   {QUERY 
   :commandType 2 
   :querySource 0 
   :canSetTag true 
   :utilityStmt <> 
   :resultRelation 2 
   :hasAggs false 
   :hasWindowFuncs false 
   :hasTargetSRFs false 
   :hasSubLinks false 
   :hasDistinctOn false 
   :hasRecursive false 
   :hasModifyingCTE false 
   :hasForUpdate false 
   :hasRowSecurity false 
   :cteList <> 
   :rtable (
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname small_view 
         :colnames ("id" "tenant_id")
         }
      :rtekind 0 
      :relid 16774 
      :relkind v 
      :rellockmode 3 
      :tablesample <> 
      :lateral false 
      :inh false 
      :inFromCl false 
      :requiredPerms 4 
      :checkAsUser 0 
      :selectedCols (b)
      :insertedCols (b)
      :updatedCols (b 8)
      :extraUpdatedCols (b)
      :securityQuals <>
      }
      {RTE 
      :alias <> 
      :eref 
         {ALIAS 
         :aliasname small 
         :colnames ("id" "tenant_id")
         }
      :rtekind 0 
      :relid 16769 
      :relkind r 
      :rellockmode 3 
      :tablesample <> 
      :lateral false 
      :inh false 
      :inFromCl true 
      :requiredPerms 4 
      :checkAsUser 10 
      :selectedCols (b 8 9)
      :insertedCols (b)
      :updatedCols (b 8)
      :extraUpdatedCols (b)
      :securityQuals <>
      }
   )
   :jointree 
      {FROMEXPR 
      :fromlist (
         {RANGETBLREF 
         :rtindex 2
         }
      )
      :quals <>
      }
   :targetList (
      {TARGETENTRY 
      :expr 
         {CONST 
         :consttype 23 
         :consttypmod -1 
         :constcollid 0 
         :constlen 4 
         :constbyval true 
         :constisnull false 
         :location 25 
         :constvalue 4 [ 1 0 0 0 0 0 0 0 ]
         }
      :resno 1 
      :resname id 
      :ressortgroupref 0 
      :resorigtbl 0 
      :resorigcol 0 
      :resjunk false
      }
      {TARGETENTRY 
      :expr 
         {VAR 
         :varno 2 
         :varattno -1 
         :vartype 27 
         :vartypmod -1 
         :varcollid 0 
         :varlevelsup 0 
         :varnoold 2 
         :varoattno -1 
         :location -1
         }
      :resno 2 
      :resname ctid 
      :ressortgroupref 0 
      :resorigtbl 0 
      :resorigcol 0 
      :resjunk true
      }
   )
   :override 0 
   :onConflict <> 
   :returningList <> 
   :groupClause <> 
   :groupingSets <> 
   :havingQual <> 
   :windowClause <> 
   :distinctClause <> 
   :sortClause <> 
   :limitOffset <> 
   :limitCount <> 
   :rowMarks <> 
   :setOperations <> 
   :constraintDeps <> 
   :withCheckOptions <> 
   :stmt_location 0 
   :stmt_len 26
   }
```

Reply:
Thanks for sharing this, so the summary is that after postgresql's rewrite (which happens before `distributed_planner` is called), the view is not on the `targetRelation`. 

What other options do we have:

1) Why do we prevent when target relation is a view? It turns out that the issue is on deparsing of the query. Can we fix that? We need to consider some extra things, so probably not easy.

2) The magic happens on postgresql's `rewriteTargetView` function. What I see is that, the function doesn't remove the view from the `rtable`, but the range table not used in `jointree`. Can we change the code such that we only error out if there is a `rtable` which is a view and it doesn't show-up in the `jointree`?  

3) This is probably easier than 2. So, Postgres only rewrites target views when the view is created based on a single table (see `view_query_is_auto_updatable`). So, can't we simply check if the `list_length(rtable) == 2` and `linitial(rtable) is a view` and `jointree doesn't have reference to 1` -- see `ExtractRangeTableIndexWalker`. So, basically, move the check for update/delete on the view from the loop, and have a separate check.

Well, 2 and 3 heavily depends on postgresql internals, but these seems very rare to change (as commented in Postgres code these are SQL standard checks). So, I'm OK with implementing 3 with some sufficient comments.

Reply:
Actually, I implemented a similar solution to 3 which relies in e844b596e45fed88fcef37fd64540adddaadbcb3. But, according to postgresql docs, it is legal for update to take many tables as argument    
`UPDATE [ ONLY ] table [ * ] [ [ AS ] alias ]`-. Then implementing 3 becomes a bit harder than we expected, if I am not wrong.

Having not much context on 1, when I observed the problem in https://github.com/citusdata/citus/pull/3104#discussion_r336479086, I initially thought that "can implementing view support be less complex than un-supporting it ?"

Reply:
I don't think you can update multiple tables with a single command in PostgreSQL. What you're referring is something else:

> table_name
The name (optionally schema-qualified) of the table to update. If ONLY is specified before the table name, matching rows are updated in the named table only. If ONLY is not specified, matching rows are also updated in any tables inheriting from the named table. Optionally, * can be specified after the table name to explicitly indicate that descendant tables are included.

well, its good to remember. We should also consider testing with partitioned tables which are inherited internally :) 


So, the commit you mention is similar to what I've in mind. Did you have any issues with that?

Reply:
Now it makes sense and I have no issues with that. So I will implement the solution 3. Then I will add the missing tests as well

Reply:
To summarize: 

We really couldn't find a robust way of detecting the `targetRelation` when it is a view. Because, the view rewriting is happening before Citus kicks in.

Our proposed solution (` if rtable[0] is a view AND view.inFromCl == false: throw error`) would be heuristic. Reading through the postgres code, I cannot see any counter examples. The target relation is always added to the `rtable` as the first item. Similarly, the relation that is replaced in place of the view is appended to the `rtable`, so it is the `second` item.

(The more intuitive approach that we considered was to trace back from the `targetRelation` to the view. But, that's also a heuristic and our proposal seems a lot simpler and looks like very similar to what this would do.)

The worst case scenario is that the query on the shards would fail, because of [this](https://github.com/citusdata/citus/issues/2046#issuecomment-465136620) explanation.

That said, happy to get any input here.

Reply:
The third solution we mentioned in https://github.com/citusdata/citus/pull/3104#discussion_r336999409 is implemented in commit 0847cce

Reply:
The third solution we mentioned in https://github.com/citusdata/citus/pull/3104#discussion_r336999409 is implemented in commit d72f17f